### DISCUSSION TITLE
Add mysql-connector-python to CI

### DISCUSSION THREAD

**Discussion by gordthompson** [sqlalchemy/sqlalchemy#11549] [Ideas]
I see quite a few SQLAlchemy-related questions (SQLA, pandas, etc.) on Stack Overflow tagged with `mysql-connector-python` or using `mysql+mysqlconnector://`.

https://www.pepy.tech/projects/mysql-connector-python

says that it gets 5M downloads per week. 

I ran the SQLA test suite using `mysql+mysqlconnector://` and only one test failed, and that is apparently due to my Docker environment because `mysql+pymysql://` also failed the same test.

Some people will prefer `mysql+mysqlconnector://` because `.executemany()` of INSERT statements is much faster than `mysql+pymysql://`. (It does "insertmanyvalues" under the hood.)

**Comment by zzzeek** [comment]
sure we can turn it back on.   you can do it straight in the tox.ini file and it will work.

just so you know, mariadbconnector and mysqlconnector have both been turned on and off in CI more than once as both drivers were not able to keep their APIs stable and kept having lots of regressions.   searching git logs most of that drama seems to be 6+ years ago.

hoping that mysqlconnector is in better shape these days, we just can't really afford to be bogged down with CI breaking because of drivers that introduce regressions  https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5390

**Reply by CaselIT** [reply]
We start on the right foot. It can't even connect in the CI:
```
INTERNALERROR> sqlalchemy.exc.DatabaseError: (mysql.connector.errors.DatabaseError) 1273 (HY000): Unknown collation: 'utf8mb4_0900_ai_ci'
```

**Reply by zzzeek** [reply]
is that because it's trying to connect to mariadbs ?

otherwise yeah.  mysql-connector.  woop

**Reply by zzzeek** [reply]
I would guess from the stack trace it has a hardcoded collation name inside of it.   it's forcing the connection collation to a fixed value?  classy

@gordthompson feel free to look into whatever is going on there.

**Reply by gordthompson** [reply]
Test suite passes for me using mysql-connector-python 9.0.0 (released this morning) against MySQL server 8.0.34 when I use

```
mysql+mysqlconnector://scott:tiger@192.168.0.199/test?charset=utf8mb4&collation=utf8mb4_unicode_ci
```

**Reply by zzzeek** [reply]
yah......we run against both mysql and mariadb server.  i think it's poor form mysql-connector wont work on mariadb server?  is there a config option?

**Reply by gordthompson** [reply]
Hmm, yes, mysql-connector-python doesn't get along nearly as well with MariaDB (version 11.0.2, anyway). It does connect, but about 85 "test suite" tests fail.

>  i think it's poor form mysql-connector wont work on mariadb server?

As we (as a team) have discussed, it seems that MariaDB and MySQL continue to go their separate ways. Also, I don't really remember MySQL making any effort to co-exist with MariaDB, it was always MariaDB that tried to be a drop-in replacement for MySQL.

> is there a config option?

Not exactly sure what you mean. Certainly nothing explicitly related to MariaDB. In fact, the mysql-connector-python documentation makes no mention of MariaDB at all. (I'm not surprised, TBH.)

The only real reason I made this suggestion is that mysql-connector-python kicks pymysql's butt when doing an upsert like

```python
insert_stmt = insert(Thing)
on_duplicate_stmt = insert_stmt.on_duplicate_key_update(
    dict(txt=insert_stmt.inserted.txt)
```

With pymysql I get

> 4,000 rows upserted in 53.6 seconds

which is 75 rows/sec. With mysql-connector-python I get

> 4,000 rows upserted in 1.2 seconds

or just over 3300 rows/sec.

**Reply by zzzeek** [reply]
well the usual client to use is mysqlclient.  works with mariadb and mysql and uses the native client.   should be as fast or faster than mysql-connector-python (I say faster because I get a strong sense mcp is overengineered)

**Reply by zzzeek** [reply]
> > is there a config option?
> 
> Not exactly sure what you mean. Certainly nothing explicitly related to MariaDB. In fact, the mysql-connector-python documentation makes no mention of MariaDB at all. (I'm not surprised, TBH.)

I mean, mysqlconnector.connect(user, pass, host, collation="..."), that is, can we make it not use "utf8mb4_0900_ai_ci'" collation or is that hardcoded into it

**Reply by gordthompson** [reply]
> well the usual client to use is mysqlclient. works with mariadb and mysql and uses the native client. should be as fast or faster than mysql-connector-python (I say faster because I get a strong sense mcp is overengineered)

mysqlclient can be a PITA to install, especially on Windows. Yes, pymysql is often slower than mysqlclient, but not 44 times slower (as would be the case if mysqlclient was at least as fast as mysql-connector-python). The difference is that pymysql (and, I suspect, mysqlclient) don't do insertmanyvalues, at least for the `INSERT … ON DUPLICATE KEY` cited above. Not sure why, as insertmanyvalues is mentioned in the dialect code.

**Reply by gordthompson** [reply]
> > > is there a config option?
> > 
> > 
> > Not exactly sure what you mean. Certainly nothing explicitly related to MariaDB. In fact, the mysql-connector-python documentation makes no mention of MariaDB at all. (I'm not surprised, TBH.)
> 
> I mean, mysqlconnector.connect(user, pass, host, collation="..."), that is, can we make it not use "utf8mb4_0900_ai_ci'" collation or is that hardcoded into it

That's what I did in the connection URL [up here](https://github.com/sqlalchemy/sqlalchemy/discussions/11549#discussioncomment-9926832).

**Reply by gordthompson** [reply]
> mysqlclient … should be as fast or faster than mysql-connector-python

Testing on Ubuntu 24.04, upserting 4000 rows, average of 3 runs.

mysql+mysqldb: 31.7 seconds, 126 rows/sec
mysql+pymysql: 25.7 seconds, 156 rows/sec
mysql+mysqlconnector: 1.2 seconds, 3333 rows/sec

YMMV, of course.

**Reply by zzzeek** [reply]
and the idea is that they are doing a SQL rewrite behind their straight executemany()?  what about using our own insertmanyvalues feature?  there shouldnt be much difference between drivers in that case.

**Reply by zzzeek** [reply]
Anyway, we can't get it to run on CI.  We are using this URL:

mariadb+mysqlconnector://scott:***@mariadb111/test?charset=utf8

and getting this:

_mysql_connector.MySQLInterfaceError: Unknown collation: 'utf8mb4_0900_ai_ci'

however, this does not happen locally.   which makes me think this is in the connector C library maybe

**Reply by gordthompson** [reply]
`use_insertmanyvalues=True` appears to have no effect in this case:

```python
engine = create_engine(
    "mysql+mysqldb://scott:tiger@192.168.0.199/test",
    use_insertmanyvalues=True,
)
```

results in

```
INSERT INTO thing (id, txt) VALUES (%s, %s) AS new ON DUPLICATE KEY UPDATE txt = new.txt
[no key 0.00027s] [(0, 'txt_0_updated'), (1, 'txt_1_updated'), (2, 'txt_2_updated'), (3, 'txt_3_updated')]
```

**Reply by zzzeek** [reply]
here is the option I was asking about

https://stackoverflow.com/questions/62829879/error-1273-but-my-schema-dont-contain-utf8mb4-0900-ai-ci

**Reply by zzzeek** [reply]
> `use_insertmanyvalues=True` appears to have no effect in this case:
> 
> ```python
> engine = create_engine(
>     "mysql+mysqldb://scott:tiger@192.168.0.199/test",
>     use_insertmanyvalues=True,
> )
> ```
> 
> results in
> 
> ```
> INSERT INTO thing (id, txt) VALUES (%s, %s) AS new ON DUPLICATE KEY UPDATE txt = new.txt
> [no key 0.00027s] [(0, 'txt_0_updated'), (1, 'txt_1_updated'), (2, 'txt_2_updated'), (3, 'txt_3_updated')]
> ```

well straight insertmanyvalues is not used for upserts.

try the approach at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-upsert-statements

**Reply by CaselIT** [reply]
> > mysqlclient … should be as fast or faster than mysql-connector-python
> 
> Testing on Ubuntu 24.04, upserting 4000 rows, average of 3 runs.
> 
> mysql+mysqldb: 31.7 seconds, 126 rows/sec 
> mysql+pymysql: 25.7 seconds, 156 rows/sec
> mysql+mysqlconnector: 1.2 seconds, 3333 rows/sec
> 
> YMMV, of course.

is there a similar difference also when using mariadb connector?

**Reply by gordthompson** [reply]
> > > mysqlclient … should be as fast or faster than mysql-connector-python
> > 
> > 
> > Testing on Ubuntu 24.04, upserting 4000 rows, average of 3 runs.
> > mysql+mysqldb: 31.7 seconds, 126 rows/sec
> > mysql+pymysql: 25.7 seconds, 156 rows/sec
> > mysql+mysqlconnector: 1.2 seconds, 3333 rows/sec
> > YMMV, of course.
> 
> is there a similar difference also when using mariadb connector?

Yes. They both clock in at around 1 second.

**Reply by gordthompson** [reply]
> try the approach at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-upsert-statements

Ah, okay. We need to specify the `.values()` clause ourselves. Thanks.

**Reply by zzzeek** [reply]
OK I guess I'll close the gerrit then ?   Here's an example of the kind of fun mysql-connector brings, this mostly confirms for me it's not worth testing:

https://jenkins.sqlalchemy.org/job/sqlalchemy_21_gerrit/cext=cext,db=mysql,pyv=py312/618/testReport/junit/test.dialect.mysql.test_for_update/MySQLForUpdateLockingTest_mariadb+mysqlconnector_11_1_2/test_joined_lock_no_subquery/

```
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jenkins/workspace/sqlalchemy_21_gerrit/eb64487a/sqlalchemy/sqlalchemy/.tox/py312-cext-mysql/lib/python3.12/site-packages/mysql/connector/connection_cext.py", line 713, in cmd_query
    raise get_mysql_exception(
sqlalchemy.exc.DatabaseError: (mysql.connector.errors.DatabaseError) 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
[SQL: UPDATE a SET x=%(x)s, y=%(y)s]
[parameters: {'x': 15, 'y': 19}]
(Background on this error at: https://sqlalche.me/e/21/4xp6)


```

**Reply by gordthompson** [reply]
> OK I guess I'll close the gerrit then ?

Sure. Sorry for the inconvenience.
